src/python/pants/backend/shell/dependency_inference.py

```diff
@@ -0,0 +1,179 @@
+async def map_shell_files() -> ShellMapping:
+    all_expanded_targets = await Get(Targets, AddressSpecs([DescendantAddresses("")]))
+    shell_tgts = tuple(tgt for tgt in all_expanded_targets if tgt.has_field(ShellSources))
+    sources_per_target = await MultiGet(
+        Get(SourcesPaths, SourcesPathsRequest(tgt[ShellSources])) for tgt in shell_tgts
+    )
+
+    files_to_addresses: dict[str, Address] = {}
+    files_with_multiple_owners: DefaultDict[str, set[Address]] = defaultdict(set)
+    for tgt, sources in zip(shell_tgts, sources_per_target):
+        for f in sources.files:
+            if f in files_to_addresses:
+                files_with_multiple_owners[f].update({files_to_addresses[f], tgt.address})
+            else:
+                files_to_addresses[f] = tgt.address
+
+    # Remove files with ambiguous owners.
+    for ambiguous_f in files_with_multiple_owners:
+        files_to_addresses.pop(ambiguous_f)
+
+    return ShellMapping(
+        mapping=FrozenDict(sorted(files_to_addresses.items())),
+        ambiguous_modules=FrozenDict(
+            (k, tuple(sorted(v))) for k, v in sorted(files_with_multiple_owners.items())
+        ),
+    )
+
+
+class ParsedShellImports(DeduplicatedCollection):
+    sort_input = True
+
+
+@dataclass(frozen=True)
+class ParseShellImportsRequest:
+    # NB: We parse per-file, rather than per-target. This is necessary so that we can have each
+    # file in complete isolation without its sibling files present so that Shellcheck errors when
+    # trying to source a sibling file, which then allows us to extract that path.
+    digest: Digest
+    fp: str
+
+
+PATH_FROM_SHELLCHECK_ERROR = re.compile(r"Not following: (.+) was not specified as input")
+
+
+@rule
+async def parse_shell_imports(
+    request: ParseShellImportsRequest, shellcheck: Shellcheck
+) -> ParsedShellImports:
+    # We use Shellcheck to parse for us by running it against each file in isolation, which means
+    # that all `source` statements will error. Then, we can extract the problematic paths from the
+    # JSON output.
+    downloaded_shellcheck = await Get(
+        DownloadedExternalTool, ExternalToolRequest, shellcheck.get_request(Platform.current)
+    )
+    input_digest = await Get(Digest, MergeDigests([request.digest, downloaded_shellcheck.digest]))
+    process_result = await Get(
+        FallibleProcessResult,
+        Process(
+            # NB: We do not load up `[shellcheck].{args,config}` because it would risk breaking
+            # determinism of dependency inference in an unexpected way.
+            [downloaded_shellcheck.exe, "--format=json", request.fp],
+            input_digest=input_digest,
+            description=f"Detect Shell imports for {request.fp}",
+            level=LogLevel.DEBUG,
+        ),
+    )
+
+    try:
+        output = json.loads(process_result.stdout)
+    except json.JSONDecodeError:
+        return ParsedShellImports()
```

Comment:
This seems worth a warning of some sort. Our expectations of shellcheck are now broken and so is the feature. This could go on silently for a while before someone reports the problem.